### PR TITLE
[mlir] [Dialect] Add LibC dialect and printf op for CPU

### PR DESCRIPTION
**Pull request by Menooker**
Add a new dialect `LibC` to expose some useful features from CPU's standard C library. Currently, only `printf` is added. `printf` is particularly useful in debugging and testing. There is a `gpu.printf` op, but it is missing on CPU.

RFC discussion at https://discourse.llvm.org/t/rfc-new-dialect-to-expose-handy-utilities/79041

### CODE REVIEW

**Comment by Menooker** [comment]
Maybe we can remove the some of the uses of `CRunnerUtils` in the future, when MLIR cpu-runner can access the libc functions?

File: mlir/include/mlir/Dialect/LibC/LibC.td [hunk 1]
```diff
@@ -0,0 +1,55 @@
+//===-- AMX.td - AMX dialect operation definitions *- tablegen -*----------===//
```
Comment (rengolin) [inline mlir/include/mlir/Dialect/LibC/LibC.td:1]: Typo

File: mlir/include/mlir/Dialect/LibC/LibC.td [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+class LibC_Op<string mnemonic, list<Trait> traits = []> :
+  Op<LibC_Dialect, mnemonic, traits> {}
+
+def LibC_PrintfOp : LibC_Op<"printf", [MemoryEffects<[MemWrite]>]>,
+  Arguments<(ins StrAttr:$format,
+                Variadic<AnyTypeOf<[AnyInteger, Index, AnyFloat]>>:$args)> {
+  let summary = "C-style printf";
+  let description = [{
+    `cpuruntime.printf` takes a literal format string `format` and an arbitrary
```
Comment (rengolin) [inline mlir/include/mlir/Dialect/LibC/LibC.td:43]:
```suggestion
    `libc.printf` takes a literal format string `format` and an arbitrary
```

**Review by rengolin — COMMENTED** [review]
FWIW, we already have some [printing primitives](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/ExecutionEngine/RunnerUtils.h).

I don't mind adding `printf`, but adding a whole dialect just for this opens a dialect discussion that is far wider than this one function.

* Can we add this to the existing primitives?
* Do we need a whole `libc` dialect? Perhaps c++ too?
* Will this help Clang with CIL? Do they already have something like this?

I'd suggest you try the easy route with the current primitives and see what that gets you.

Some nits on the code. Also, your tests are not enough. You need to cover the functionality you add, including the error messages that you expect to get on a _negative_ test.

**Comment by Menooker** [comment]
Thanks for @rengolin for the view!

> Can we add this to the existing primitives?

What I would like to do is to add the functionality of printing values in MLIR. It would be very helpful when debugging and writing tests, since users can print arithmetic values any where in the MLIR code. Also developers can use filecheck on the outputs by printf. The printing primitives you mentioned does not support formatting and need to declare the functions before using, which is not easy to use in IR.

> Do we need a whole libc dialect? Perhaps c++ too?

Other interesting features in libc I know will be the ability of `assert` and `abort`. We may introduce them as Ops in `LibC` dialect too. `assert` is a C-macro, but I think it is a good-to-have in MLIR, for example, to help debug the program in "-O0" mode, with printing the `Source Location` . Based on it, some other features can be implemened. Memref dialect can use `libc.assert` to make sure the user-passed memref in func args have the claimed rank and dimensions at the entry. Another useful feature based on `libc.assert` is the boundary checking on memory accesses on memref. I am not sure if these features are already in MLIR or not.

C++ has complex ABI issues and many of its useful features are based on templates. I don't think we need `libcpp` dialect for now.

>  adding a whole dialect just for this opens a dialect discussion that is far wider than this one function.

I totally understand that it is overkill to add a dialect for a single Op. My original proposal is to add an printf op for CPU. I have found an `gpu.printf`. There is no reason not having a CPU one, since it is easier to implement and equally useful. However, there is no such `cpu` dialect to hold the `printf` op, and there seems nowhere to place such op. That's why I added this dialect.

After more searching, I found that in `vector` dialect, there is `vector.print` which can print scalar values and etc. And this op lowers to `printF32` or other functions in `CRunnerUtils` on CPU. So there may be 3 ways for me to introduce this printf for CPU:

1. Add to vector dialect, as `vector.printf`. However, it seems a bit strange to put such op in vector dialect.
2. Instead of introducing `libc` dialect, introduce a new `print` dialect. Besides `print.printf`, we can introduce `print.printMemref` (`print.printTensor` also?)
3. Introduce `libc` dialect. Add more Ops into it, like `abort` and `assert` discussed above. Or we may call it `debug` dialect, for debugging purpose ops?

> Some nits on the code. Also, your tests are not enough. You need to cover the functionality you add, including the error messages that you expect to get on a negative test.

Thanks for pointing out! I am an MLIR beginner here. Will improve them! :)

**Comment by rengolin** [comment]
> The printing primitives you mentioned does not support formatting and need to declare the functions before using, which is not easy to use in IR.

Indeed, we had the same problem in the past, but decided to [print tensors/memrefs as vectors](https://github.com/plaidml/tpp-mlir/blob/08caaf675eed621f6ae06c83bc7bd06441550bf8/lib/TPP/Runner/MLIRBench.cpp#L318).

> Other interesting features in libc I know will be the ability of `assert` and `abort`. We may introduce them as Ops in `LibC` dialect too. `assert` is a C-macro, but I think it is a good-to-have in MLIR, for example, to help debug the program in "-O0" mode, with printing the `Source Location` . Based on it, some other features can be implemened. Memref dialect can use `libc.assert` to make sure the user-passed memref in func args have the claimed rank and dimensions at the entry. Another useful feature based on `libc.assert` is the boundary checking on memory accesses on memref. I am not sure if these features are already in MLIR or not.

We have added "runner" functionality in local dialects ([check](https://github.com/plaidml/tpp-mlir/blob/main/include/TPP/Dialect/Check/CheckOps.td), [perf](https://github.com/plaidml/tpp-mlir/blob/main/include/TPP/Dialect/Perf/PerfTypes.td)), but never quite found the place to put them. 

I think there's a common trend between all these little functions and it's not `libc`. I'd call it `runner` or `runtime` or just `run` and de-compose the `gpu.printf` into `gpu.get_tensor` + `run.printf`, for example. I'd also add the benchmark loops, checking of values, etc.

> C++ has complex ABI issues and many of its useful features are based on templates. I don't think we need `libcpp` dialect for now.

Right, I was using it to show how quickly this can escalate if we chose a tangent path. :)

I do like the idea of improving the ability to print stuff in MLIR, but just a printf wrapper is too much for a new dialect.

Right now, you can already lower your formatted printf into LLVM dialect call, so this should be enough for _just_ basic printf. On the rest of the _"runtime dialect"_, I think an RFC is in order to make sure we converge to a common design.

I'd love to find a home for our `check` and `perf` dialects.

**Comment by joker-eph** [comment]
Thanks for the contribution! Please have a look at https://mlir.llvm.org/getting_started/DeveloperGuide/#guidelines-on-contributing-a-new-dialect-or-important-components  and conduct a discussion on Discourse for contributing a new dialect.

**Comment by Menooker** [comment]
> Thanks for the contribution! Please have a look at https://mlir.llvm.org/getting_started/DeveloperGuide/#guidelines-on-contributing-a-new-dialect-or-important-components and conduct a discussion on Discourse for contributing a new dialect.

Thanks for reminder. I have posted an [RFC](https://discourse.llvm.org/t/rfc-new-runtime-dialect/79041). We can move to that for discussion on the new dialect.

**Pull request closed without merging**